tokio/src/macros/select.rs

```diff
@@ -0,0 +1,788 @@
+/// Wait on multiple concurrent branches, returning when the **first** branch
+/// completes, cancelling the remaining branches.
+///
+/// The `select` macro accepts one or more branches with the following pattern:
+///
+/// ```text
+/// <pattern> = <async expression> (, if <condition>)? => <handler>,
+/// ```
+///
+/// Additionally, the `select!` macro may include a single, optional `else`
+/// branch, which evaluates if none of the other branches match their patterns:
+///
+/// ```text
+/// else <expression>
+/// ```
+///
+/// The macro aggregates all `<async expression>` expressions and runs them
+/// concurrently on the **current** task. Once the **first** expression
+/// completes with a value that matches its `<pattern>`, the `select!` macro
+/// returns the result of evaluating the completed branch's `<handler>`
+/// expression.
+///
+/// Additionally, each branch may include an optional `if` condition. This
+/// condition is evaluated **before** the <async expression>. If the condition
+/// returns `false`, the branch is entirely disabled. This capability is useful
+/// when using `select!` within a loop.
+///
+/// The complete lifecycle of a `select!` expression is as follows:
+///
+/// 1. Aggregate the `<async expression>`s from each branch.
```

Comment:
Should this happen for disabled branches? I actually expected step 1) and 2) to be swapped

Reply:
The async expression is evaluated first, the condition second. The future returned from the async expression is **not** polled unless the condition returns true.